buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesTask.java

```diff
@@ -0,0 +1,126 @@
+            ) {
+                dependencyError(format("modules may not depend on launcher modules. Invalid dependency: %s", dependency));
+            }
+
+            if (pathFromRoot.startsWith("samples/") && // no module may depend on samples (exceptions follow)
+                    !project.getPath().startsWith(":samples:") // exception: other samples might depend on samples
+            ) {
+                dependencyError(format("modules may not depend on samples modules. Invalid dependency: %s", dependency));
+            }
+
+            if (pathFromRoot.startsWith("system-tests/") // no module may depend on system-tests
+            ) {
+                dependencyError(format("modules may not depend on system-tests modules. Invalid dependency: %s", dependency));
+            }
+
+            if (CROSS_MODULE_DEPENDENCY_PATTERN.matcher(pathFromThisModule.toString()).matches() && // there should not be "cross-module" dependencies at the same level
+                    !dependency.getName().endsWith("-core") && // exception: technology libs such as "blob-core"
+                    !pathFromRoot.startsWith("samples/") // exception: samples may refer to own modules
+            ) {
+                dependencyError(format("there should not be \"cross-module\" dependencies at the same level. Invalid dependency: %s", dependency));
+            }
+
+            if (project.getName().equals("core-spi")) { // `core:spi` cannot depend on any other module
+                dependencyError(format("core:spi` cannot depend on any other module. Invalid dependency: %s", dependency));
+            }
+
+            if (project.getName().equals(dependency.getName())) { // two modules cannot have the same name (TBC)
+                dependencyError(format("two modules cannot have the same name. Invalid dependency: %s", dependency));
+            }
+        }
+    }
+
+    private void directDependencyRules(ResolvedDependency dependency) {
+        var project = this.getProject();
+
+        if (dependency.getModuleGroup().equals(project.getGroup())) {
+            dependency.getModuleArtifacts().forEach(artifact -> {
+                var pathFromRoot = project.getRootDir().toPath().relativize(artifact.getFile().toPath());
+                if (pathFromRoot.startsWith("core/") && // no module may depend directly on any core module
+                        !dependency.getName().endsWith("-core") // exception: other core modules
+                ) {
+                    dependencyError(format("modules may not depend directly on core modules. Invalid dependency: %s", dependency.getName()));
+                }
+            });
+        }
+    }
+
+    private void dependencyError(String error) {
+        var message = format("DEPENDENCY RULE VIOLATION: %s", error);
+        if (getFailOnError().get()) {
+            throw new GradleException(message);
+        } else {
+            System.out.println(message);
```

Comment:
is `System.out.println` this the recommended way to show output in a plugin?

Reply:
Looking at samples in the documentation it seems that is the way indeed. This is analog to Kotlin's 'println' as well.

Reply:
I found that is possible to use a logger as well. Switching from `System.out` to the Task's logger.